examples/larger-examples/datalog/subtyping.flix

```diff
@@ -0,0 +1,301 @@
+mod Subtyping {
+
+    type alias Program = {
+        classes = Vector[String],
+        interfaces = Vector[String],
+        classImplements = Vector[(String, String)],
+        classExtends = Vector[(String, String)],
+        interfaceExtends = Vector[(String, String)],
+        fields = Vector[(String, String)],
+        classMethods = Vector[(String, String)],
+        interfaceSignatures = Vector[(String, String)]
+    }
+
+    def classHierarchyAnalysis(p: Program): #{ Subtype(String, String), AvailableField(String, String), AvailableMethod(String, String), NonImplementedMethod(String, String) | r } =
+        let cs    = inject p.classes             into Class;
+        let is    = inject p.interfaces          into Interface;
+        let impls = inject p.classImplements     into ClassImplements;
+        let ce    = inject p.classExtends        into ClassExtends;
+        let ies   = inject p.interfaceExtends    into InterfaceExtends;
+        let fs    = inject p.fields              into Field;
+        let ms    = inject p.classMethods        into ClassMethod;
+        let iss   = inject p.interfaceSignatures into InterfaceSignature;
+        let rules = #{
+            Subtype(x, x)       :- Class(x).
+            Subtype(x, x)       :- Interface(x).
+            Subtype(sub, super) :- Class(sub), Interface(super), ClassImplements(sub, super).
+            Subtype(sub, super) :- Class(sub), Class(super), ClassExtends(sub, super).
+            Subtype(sub, super) :- Interface(sub), Interface(super), InterfaceExtends(sub, super).
+            Subtype(sub, super) :- Subtype(sub, a), Subtype(a, super).
+
+            AvailableField(class, field)        :- Field(t, field), Subtype(class, t).
+            AvailableMethod(class, method)      :- ClassMethod(t, method), Subtype(class, t).
```

Comment:
The (new) idea is that `AvailableField` and `AvailableMethod` captures fields/methods which must be available on a type. But since a type can be abstract (i.e. interface or class) then the programmer may have forgot to specify a method. We capture the methods _actually_ available in the ConcreteMethod relation:

```
ConcreteMethod(class, method) :- Class(class), ClassMethod(class, method).
ConcreteMethod(sub, method) :- Class(sub), Subtype(sub, super), ConcreteMethod(super, method).
```

We expose `(Subtype, AvailableField, AvailableMethod, ConcreteMethod)` as the return type. Drop  `NonImplementedMethod`.

Later, before `illegalFinalClasses`, we can add something like:

```
def missingMethods(...): Vector[...] = 
  inject p.interfaces into Interface;
 ... #{  MissingMethod(class, method) :- AvailableMethod(class, method), not ConcreteMethod(class, method), not Interface(class).
     }
// Note: Here you have to inject "Interface" _again_, but that is acceptable. 
```